routers/api/packages/arch/arch.go

```diff
@@ -0,0 +1,204 @@
+		Distribution: distro,
+		Buffer:       buf,
+	})
+	if err != nil {
+		apiError(ctx, http.StatusBadRequest, err)
+		return
+	}
+
+	_, err = buf.Seek(0, io.SeekStart)
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+
+	props := map[string]string{
+		distro + "-" + filename + ".desc": desc.String(),
+	}
+	if sign != "" {
+		_, err := hex.DecodeString(sign)
+		if err != nil {
+			apiError(ctx, http.StatusBadRequest, err)
+		}
+		props[distro+"-"+filename+".sig"] = sign
+	}
+
+	ver, _, err := pkg_service.CreatePackageOrAddFileToExisting(
+		&pkg_service.PackageCreationInfo{
+			PackageInfo: pkg_service.PackageInfo{
+				Owner:       ctx.Package.Owner,
+				PackageType: pkg_model.TypeArch,
+				Name:        desc.Name,
+				Version:     desc.Version,
+			},
+			Creator: ctx.ContextUser,
+			Metadata: &arch_module.Metadata{
+				URL:          desc.URL,
+				Description:  desc.Description,
+				Provides:     desc.Provides,
+				License:      desc.License,
+				Depends:      desc.Depends,
+				OptDepends:   desc.OptDepends,
+				MakeDepends:  desc.MakeDepends,
+				CheckDepends: desc.CheckDepends,
+				Backup:       desc.Backup,
+			},
+		},
+		&pkg_service.PackageFileCreationInfo{
+			PackageFileInfo: pkg_service.PackageFileInfo{
+				Filename:     filename,
+				CompositeKey: distro + "-" + filename,
+			},
+			OverwriteExisting: true,
+			IsLead:            true,
+			Creator:           ctx.ContextUser,
+			Data:              buf,
+			Properties:        props,
+		},
+	)
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = arch_service.RepoConnect(ctx, owner, desc.Name, ver.ID)
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+
+	ctx.Status(http.StatusOK)
+}
+
+// Get file from arch package registry.
+func Get(ctx *context.Context) {
+	var (
+		file   = ctx.Params("file")
+		owner  = ctx.Params("username")
+		distro = ctx.Params("distro")
+		arch   = ctx.Params("arch")
+	)
+
+	if strings.HasSuffix(file, ".pkg.tar.zst") {
+		pkg, err := arch_service.GetPackageFile(ctx, distro, file)
+		if err != nil {
+			apiError(ctx, http.StatusNotFound, err)
+			return
+		}
+
+		ctx.ServeContent(pkg, &context.ServeHeaderOptions{
+			Filename: file,
+		})
+		return
+	}
+
+	if strings.HasSuffix(file, ".pkg.tar.zst.sig") {
+		sig, err := arch_service.GetPackageSignature(ctx, distro, file)
+		if err != nil {
+			apiError(ctx, http.StatusNotFound, err)
+			return
+		}
+
+		ctx.ServeContent(sig, &context.ServeHeaderOptions{
+			Filename: file,
+		})
+		return
+	}
+
+	if strings.HasSuffix(file, ".db.tar.gz") || strings.HasSuffix(file, ".db") {
+		db, err := arch_service.CreatePacmanDb(ctx, &arch_service.DbParams{
+			Owner:        owner,
+			Architecture: arch,
+			Distribution: distro,
+		})
+		if err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+			return
+		}
+
+		ctx.ServeContent(bytes.NewReader(db), &context.ServeHeaderOptions{
```

Comment:
`CreatePacmandDb` uses a buffer internally but returns `[]byte` just to create a new reader. Could just return the buffer or `io.Reader`.

Reply:
@KN4CK3R 

There is no `Seek()` method to pass `bytes.Buffer` directly to `ctx.ServeContent()`. Replaced with `HashedBuffer` from packages `module`.